src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs

```diff
@@ -0,0 +1,889 @@
+                .Select(FilteredNone)
+                .Select(FilteredInsert)
+                .Select(FilteredMove)
+                .TimeInterval()
+                .Select(UpdateProcessingDelay)
+                .Select(data => data.Item)
+                .Publish()
+                .RefCount();
+
+            return source;
+        }
+
+        /// <summary>
+        /// Set a new comparer for the existing data. This will cause the
+        /// collection to be resorted and refiltered.
+        /// </summary>
+        /// <param name="theComparer">The comparer method for sorting, or null if not sorting</param>
+        public void SetComparer(Func<T, T, int> theComparer)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            SetAndRecalculateSort(theComparer);
+            SetAndRecalculateFilter(filter);
+        }
+
+        /// <summary>
+        /// Set a new filter. This will cause the collection to be filtered
+        /// </summary>
+        /// <param name="theFilter">The new filter, or null to not have any filtering</param>
+        public void SetFilter(Func<T, int, IList<T>, bool> theFilter)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            SetAndRecalculateFilter(theFilter);
+        }
+
+        public IDisposable Subscribe()
+        {
+            if (source == null)
+                throw new InvalidOperationException("No source observable has been set. Call Listen or pass an observable to the constructor");
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            disposables.Add(source.Subscribe());
+            return this;
+        }
+
+        public IDisposable Subscribe(Action<T> onNext, Action onCompleted)
+        {
+            if (source == null)
+                throw new InvalidOperationException("No source observable has been set. Call Listen or pass an observable to the constructor");
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            disposables.Add(source.Subscribe(onNext, onCompleted));
+            return this;
+        }
+
+        public void AddItem(T item)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+            queue.Enqueue(item);
+        }
+
+        public void RemoveItem(T item)
+        {
+            if (disposed)
+                throw new ObjectDisposedException("TrackingCollection");
+
+            var position = GetIndexUnfiltered(item);
+            if (position < 0)
+                return;
+
+            var data = new ActionData(TheAction.Remove, item, null, position - 1, position, original);
+            data = CheckFilter(data);
+            data = CalculateIndexes(data);
+            data = SortedRemove(data);
+            data = FilteredRemove(data);
```

Comment:
If it's a pipeline, shouldn't each method return a new instance of `ActionData` instead of mutating the passed in instance? I think this calling pattern is really confusing.

Also, `SortedRemove` and `FilteredRemove` actually mutate the current `TrackingCollection`. They are very different from the methods called before them, which do some calculations and return an ActionData with the calculations. I think that difference and intent would be better communicated by making those two methods return `void`.

So this method would look like.

```csharp
var data = new ActionData(TheAction.Remove, item, null, position - 1, position, original);
data = CheckFilter(data);
data = CalculateIndexes(data);
SortedRemove(data);
FilteredRemove(data);
```

Where `data` is never mutated. Thoughts?